lib/README.md

```diff
@@ -57,6 +57,17 @@ Don't have multiple changes in one PR, instead split it up into multiple ones.
 
 This keeps the conversation focused and has a higher chance of getting merged.
 
+### Ensure backwards compatibility
+
+`lib` is the effective standard library for Nix, both for Nixpkgs and third-parties.
+Because of that it's important to make sure that the interface isn't changed in a breaking way.
+
+The only exceptions are:
+- Breaking changes to interfaces that aren't in a release yet
+- Breaking changes that are unlikely to cause problems for users
+- Removal of functions that have been deprecated for at least one release
```

Comment:
Does this mean it needs to be deprecated for 6 calendar months, or that it was deprecated before any previous release and functions can be removed the day after the release it was deprecated in?

Reply:
They can be removed immediately following the release branch-off, so that the release will have the warning, while the next one won't, I'll clarify this.

Reply:
I'd prefer for deprecations to start after old stable EOL, especially if it doesn't have the new alternative. That adds one month, but provides a guarantee that third party libraries work with all currently supported versions of lib, ensuring that maintainers don't get caught in this sequence of events:
- users report deprecation warning
- maintainer acts on it
- users report evaluation error on a supported Nixpkgs release
- frustrated maintainer has to revert or write a workaround

Alternatively, we could require a backporting of the new thing to stable / old stable to mitigate compatibility issues, but that's slightly more effort.